producer/src/main/kotlin/streams/procedures/StreamsProcedures.kt

```diff
@@ -38,6 +50,8 @@ class StreamsProcedures {
                 .withTopic(topic)
                 .build()
         StreamsProcedures.eventRouter.sendEvents(topic, listOf(streamsEvent))
```

Comment:
this is asynchronous, afaik 

so we're not sure if the event was actually sent, it's more like fire&forget.

Reply:
@conker84 wdyt

Reply:
Yes it's an async call. 
It would be cool to add a new method `sendSync` into StreamsEventRouter and implement it in the `KafkaEventRouter` in this way:

```kotlin
    override fun startSync(producerRecord: ProducerRecord<String, ByteArray>): RecordMetadata {
        producer.send(producerRecord).get()
    }
``` 

So we can provide as output the `RecordMetadata`

Reply:
ah you're right this is even better.  I'll look into it.

Reply:
Pushed updates.  @conker84 had implemented "fire and forget".  I changed the return types of several functions to return Futures.  The fire and forget behavior of this method has been retained, **but** I have implemented a new sendSync method according to @conker84 's suggestion.  I changed streams.publish to use this, so now it can return record metadata to the client.

"Fire and forget" is still possible with the API, and at a later point we might want to make this a config option.  Default behavior for streams.publish (in my opinion) should be to wait until it can confirm it did what you asked.  This matches the behavior of streams.consume which polls / blocks.

Reply:
I was wondering if instead of returning the `RecordMetadata`, it would be better to return a meta-structure that should be general as possible. This because we designed the Stream module in order to provide different "streaming connectors" (kinesis, rabbitMQ, ...)
@jexp @moxious wdyt?

Reply:
If you like, a solution here is I could return record metadata as a map.  In this way I can include all the kafka-specific metadata, but the return type of the procedure can always be the same.  If it's a different streaming tech, you still get recordMetadata as a result, but the map has different keys.

Reply:
Cool!

Reply:
@jexp @conker84 I'm going down a deep rabbit hole here and might need some help.

So I created a wrapper class PublishedRecordMetadata and then implemented that in a specific case for kafka to make this generic.  However the Kafka publish functions all return Future<RecordMetadata>.   Doing research, these Futures are not pipe-lineable.  (And so doing the trivial bit of turning Future<RecordMetadata> into Future<KafkaPublishedRecordMetadata> is turning out to be stupid complex)

 In the Java world they're telling me that I need a CompletableFuture because Future is the old API.  In the Kotlin world they're telling me that I should use something else entirely.

So basically this simple suggestion has turned into something much more complicated for me with Kotlin/java async programming.

What is your opinion on how this should be done?  Forgetting for a moment the RecordMetadata class, what async implementation do you want this library to return e.g. return type here:

```
fun send(producerRecord: ProducerRecord<String, ByteArray>): Future<PublishedRecordMetadata>
```

I don't want to prep the PR again and find out that it needs to be totally different . :)

Reply:
Maybe we can consider to keep it simple and change the _old_ `sendEvent` method in orderd to take a `callback` as third argument (so it can still be used with the fire-and-forget style); at this point the new `sendEventSync` should simply returns a `List<PublishedRecordMetadata>`

wdyt?

Reply:
Not sure I understand how that would work.   Suppose I took a callback, I'm still going to get a `Future<RecordMetadata>` from the Kafka API.  If I write some code to `.get()` the value from that future, it's going to block in the async method -- or else if it doesn't block, then passing a callback is really the same as just calling `.get()` on the future yourself.

Isn't it also sort of the case that adding callbacks to APIs tends to make them messier and encourage "pyramid" style programming, I think Futures and promises were intended to get around this.   It does seem like a nasty combination of Java/Kotlin issues here though.

Reply:
I think we should convert the `Future` in a `Deferred` by using (the Kotlin extension function [feature](https://kotlinlang.org/docs/reference/extensions.html)) in this way:
```kotlin
fun <T> Future<T>.asDeferred(dispatcher: CoroutineDispatcher = Dispatchers.Default): Deferred<T> {
    return GlobalScope.async(dispatcher) {
        get()
    }
}
```

An use it in that way:
```kotlin
private fun send(producerRecord: ProducerRecord<String, ByteArray>): Deferred<RecordMetadata> {
  return producer.send(producerRecord) { meta: RecordMetadata?, error: Exception? ->
    if (log.isDebugEnabled) {
      log.debug("Sent record in partition ${meta?.partition()} offset ${meta?.offset()} data ${meta?.topic()} key size ${meta?.serializedKeySize()}", error)
    }
  }.asDeferred()
}
```

Reply:
I had a chat on the Kotlin Slack and basically, we have two options:
- change the implementation of `asDeferred` in this way:
```
fun <T> Future<T>.asDeferred(): Deferred<T> {
    return GlobalScope.async(Dispatchers.IO) {
        get()
    }
}
```
- implement a solution as described [here](https://stackoverflow.com/questions/50793362/how-to-use-suspendcoroutine-to-turn-java-7-future-into-kotlin-suspending-functio#answer-52901712)